### ISSUE TITLE
broken continuation line

### ISSUE THREAD

**Issue by asartori86** [dealii/dealii#2098]
recent PR #2078 changed the behavior of the continuation line in parameter files.  Before that PR this worked for me
```
set Finite element space             = FESystem[FE_Q(1)^d-\
                                                            FE_DGPMonomial(0)-\
					                    FE_DGPMonomial(0)]
```

now I got

```
An error occurred in line <1480> of file </apps/src/dealii/source/fe/fe_tools.cc> in function
    dealii::FiniteElement<dim, spacedim>* dealii::FETools::internal::{anonymous}::get_fe_from_name_ext(std::string&, const std::map<std::basic_string<char>, std::shared_ptr<const dealii::Subscriptor> >&) [with int dim = 3; int spacedim = 3; std::string = std::basic_string<char>]
The violated condition was: 
    fe_name_map.find(name_part) != fe_name_map.end()
The name and call sequence of the exception was:
    ExcInvalidFEName(name)
Additional Information: 
Can't re-generate a finite element from the string '                                         FE_DGPMonomial(0)-      FE_DGPMonomial(0)]'.
```

**Comment by drwells** [comment]
Yes, I bet I know what happened: when I moved the code up I got rid of the extra whitespace trimming, believing it to be not necessary.

Patch incoming.

**Comment by asartori86** [comment]
thanks

**Comment by bangerth** [comment]
The solution is to trim whitespace at the end of the continued line as well as the beginning of the continuing line. This is certainly the element of least surprise. One could argue about leaving a single space  to ensure two numbers provided on successive lines are not just collated.

**Comment by drwells** [comment]
I wanted to ask for another opinion before changing the behavior in `ParameterHandler`. In particular, when `cpp` handles continuations, it does not delete whitespace at the start of lines: my concern is that we may accidentally start combining tokens with this behavior. Is this still the right thing to do?

I doubt it will matter in practice; `ParameterHandler` will probably fail in an obvious way if tokens are erroneously combined.

If we leave a single space then `get_fe_by_name` (without #2100) would not work.

**Comment by tjhei** [comment]
> The solution is to trim whitespace at the end of the continued line as well as the beginning of the continuing line. This is certainly the element of least surprise

Agreed. This is also in line (pun intended) with how we treat single line values: we strip the white space before and after.

> One could argue about leaving a single space to ensure two numbers provided on successive lines are not just collated. 

I would prefer we don't do that. If you want a space, insert it before the ``\``.

**Comment by drwells** [comment]
I thought about it some more and I believe now my concerns are misplaced. If this does cause any weird tokenization errors they should be easy to spot. How about:
1. If a line ends with something like `\__` (pretend the underscores are spaces), then trim any whitespace after the `\` and continue as usual. `cpp` does this, albeit with a warning.
2. If the current line continues the previous line, then trim all whitespace at the beginning of the current line. `cpp` does not do this, but this is probably better suits our needs.

I will put up a PR soon.

**Comment by bangerth** [comment]
> I would prefer we don't do that. If you want a space, insert it before
> the |\|.

In that case, trimming space at the beginning of the second line makes 
no sense, and it just reproduces the current behavior.

People like to align the |\| characters. So I think there will often be 
space before it. My take would be that if you rely on a space being 
somewhere, make sure it's not the one between the end of the line and 
the backslash, or between the beginning of the line and the first 
non-whitespace character. Using whitespace as a separator is poor 
practice anyway.

**Comment by asartori86** [comment]
> 1. If a line ends with something like \__ (pretend the underscores are spaces), then trim any whitespace after the \ and continue as usual. cpp does this, albeit with a warning.

In my previous PR I kept the \ as continuation line if and only if it was the last character of the line. In my opinion it is an error to put characters after the \. More precisely, the continuation line symbol in my mind is \ plus the symbol of new line (e.g. \n, $ etc..)

**Comment by drwells** [comment]
@asartori86 Does everything work for you now? I believe the patches resulting from the discussion in #2101 should fix this.

**Comment by asartori86** [comment]
@drwells , it works. Thanks

**Comment by asartori86** [comment]
Closed #2098 by #2101
